src/http/client.cc

```diff
@@ -321,37 +338,24 @@ future<> client::make_request(request&& req, reply_handler&& handle, std::option
     });
 }
 
-static bool is_retryable_exception(std::exception_ptr ex) {
-    while (ex) {
-        try {
-            std::rethrow_exception(ex);
-        } catch (const std::system_error& sys_err) {
-            auto code = sys_err.code().value();
-            if (code == EPIPE || code == ECONNABORTED || code == ECONNRESET || code == GNUTLS_E_PREMATURE_TERMINATION) {
-                return true;
-            }
-            try {
-                std::rethrow_if_nested(sys_err);
-            } catch (...) {
-                ex = std::current_exception();
-                continue;
-            }
-            return false;
-        } catch (const httpd::response_parsing_exception&) {
-            return true;
-        } catch (const std::exception& e) {
-            try {
-                std::rethrow_if_nested(e);
-            } catch (...) {
-                ex = std::current_exception();
-                continue;
-            }
-            return false;
-        } catch (...) {
-            return false;
+future<> client::maybe_retry_request(std::exception_ptr ex,
+                                     unsigned retry_count,
+                                     const request& req,
+                                     reply_handler& handle,
+                                     const retry_strategy& strategy,
+                                     std::optional<reply::status_type> expected,
+                                     abort_source* as) {
+    return strategy.should_retry(ex, retry_count).then([this, ex = std::move(ex), retry_count, &req, &handle, &strategy, as, expected](bool retry) mutable {
```

Comment:
This copies the exception.
It can be made the other way:

```
future<> stratey::should_retry(std::exception_ptr ex, ...) {
    if (yes) {
        return make_ready_future<>();
    } else {
        return make_exception_future<>(std::move(ex));
    }
}
```

and maybe_retry_request doesn't need to keep ex copy in should_retry and and continuation lambda (and doesn't need to check bool return value)

Thoughts?

Reply:
> This copies the exception.

Nope, it is like `shared_ptr` it just counts references for the owned exception

> It can be made the other way:
> 
> ```
> future<> stratey::should_retry(std::exception_ptr ex, ...) {
>     if (yes) {
>         return make_ready_future<>();
>     } else {
>         return make_exception_future<>(std::move(ex));
>     }
> }
> ```
> 
> and maybe_retry_request doesn't need to keep ex copy in should_retry and and continuation lambda (and doesn't need to check bool return value)
> 
> Thoughts?

This is how I did it initially and didnt like it, it looked really odd